### ISSUE TITLE
Epsilon Framework vulnerability

### ISSUE THREAD

**Issue by nodarik** [puikinsh/Sparkling#262]
Please check this ticket with more details: https://colorlibsupport.com/t/warning-this-theme-is-not-secure/31010

Shortly: https://www.wordfence.com/blog/2021/12/massive-wordpress-attack-campaign/

**Comment by cristianraiber** [comment]
those have been fixed a long time ago

**Comment by StudioMaX** [comment]
@cristianraiber any details please?
Because submodule of Epsilon Framework https://github.com/ColorlibHQ/Sparkling/tree/master/inc/libraries was last updated on June 30, 2017.

**Comment by cristianraiber** [comment]
As mentioned above, those vulnerabilities have been fixed a long time ago.

**Comment by StudioMaX** [comment]
Sorry for being intrusive, but I asked because I see that this vulnerability has been fixed in other Colorlib themes, for example in Shapely between versions 1.2.7 and 1.2.8. But in the latest version of Sparkling theme (2.4.8), I still see the same code fragments that were fixed in other themes. Like these changes:
```patch
--- shapely/1.2.7/inc/libraries/epsilon-framework/classes/class-epsilon-ajax-controller.php	2022-02-09 01:43:26.000000000 +0600
+++ shapely/1.2.8/inc/libraries/epsilon-framework/classes/class-epsilon-ajax-controller.php	2022-02-09 01:43:29.000000000 +0600
@@ -16,16 +16,13 @@
 		 * Action for easier AJAX handling
 		 */
 		add_action( 'wp_ajax_epsilon_framework_ajax_action', array(
 			$this,
 			'epsilon_framework_ajax_action',
 		) );
-		add_action( 'wp_ajax_nopriv_epsilon_framework_ajax_action', array(
-			$this,
-			'epsilon_framework_ajax_action',
-		) );
+
 	}
 
 	/**
 	 * Ajax handler
 	 */
 	public function epsilon_framework_ajax_action() {
@@ -50,24 +47,25 @@
 						'error'  => esc_html__( 'Not allowed', 'epsilon-framework' ),
 					)
 				)
 			);
 		}
 
-		if ( ! class_exists( $args_action[0] ) ) {
+        $class = Epsilon_Ajax_Controller::sanitize_class_name( $args_action[0] );
+
+		if (! $class || ! class_exists( $class ) ) {
 			wp_die(
 				wp_json_encode(
 					array(
 						'status' => false,
 						'error'  => esc_html__( 'Class does not exist', 'epsilon-framework' ),
 					)
 				)
 			);
 		}
 
-		$class  = $args_action[0];
 		$method = $args_action[1];
 
 		if ( 'generate_partial_section' === $method ) {
 			$args = array_map( 'Epsilon_Ajax_Controller::sanitize_arguments_for_output', wp_unslash( $_POST['args']['args'] ) );
 		} else {
 			$args = isset( $_POST['args']['args'] ) ? $_POST['args']['args'] : $_POST['args'];
@@ -111,12 +109,26 @@
 			return array_map( 'sanitize_text_field', $args );
 		} else {
 			return sanitize_text_field( $args );
 		}
 	}
 
+    /**
+     * Sanitize class name
+     *
+     * @param $args
+     */
+    public static function sanitize_class_name( $class ) {
+        $allowed_classes = array( 'Epsilon_Helper', 'Epsilon_Notify_System', 'Epsilon_Page_Generator', 'Epsilon_Typography', 'Epsilon_Color_Scheme', 'Epsilon_Notifications' );
+        if ( in_array( $class, $allowed_classes ) ) {
+            return $class;
+        }else{
+            return false;
+        }
+    }
+
 	/**
 	 * Sanitize arguments for output
 	 *
 	 * @param $args
 	 */
 	public static function sanitize_arguments_for_output( $args ) {

```

```patch
--- shapely/1.2.7/inc/libraries/welcome-screen/class-epsilon-welcome-screen.php	2022-02-09 01:43:27.000000000 +0600
+++ shapely/1.2.8/inc/libraries/welcome-screen/class-epsilon-welcome-screen.php	2022-02-09 01:43:30.000000000 +0600
@@ -154,18 +154,12 @@
 		add_action(
 			'wp_ajax_welcome_screen_ajax_callback', array(
 				$this,
 				'welcome_screen_ajax_callback',
 			)
 		);
-		add_action(
-			'wp_ajax_nopriv_welcome_screen_ajax_callback', array(
-				$this,
-				'welcome_screen_ajax_callback',
-			)
-		);
 
 		if ( $this->edd ) {
 			/**
 			 * Initiate EDD Stuff
 			 */
 			add_action( 'admin_init', array( 'EDD_Theme_Helper', 'init' ) );
```
Therefore, I just wanted to know in which version this vulnerability was fixed, since security researchers still say that the vulnerability has not been fixed in Sparkling.

**Comment by cristianraiber** [comment]
cc @razvanaldea89

**Comment by razvanaldea89** [comment]
Seems the issue is still present in the Epsilon Framework. Going to issue a fix shortly

**Comment by bjoerns1983** [comment]
Is the fix included in Sparkling 2.4.9?

**Comment by razvanaldea89** [comment]
Hello @bjoerns1983 ,

Yes, this fix was the main reason of Sparkling version 2.4.9
